app/models/extract_filter.rb

```diff
@@ -1,32 +1,75 @@
 class ExtractFilter
+  class ExtractsForClassification
+    attr_reader :extracts
+
+    def initialize(extracts)
+      @extracts = extracts
+    end
+
+    def select(&block)
+      self.class.new(extracts.select(&block))
+    end
+
+    def classification_id
+      extracts.first.classification_id
+    end
+
+    def classification_at
+      extracts.first.classification_at
+    end
+
+    def user_id
+      extracts.first.user_id
+    end
+  end
+
   attr_reader :extracts, :filters
 
   def initialize(extracts, filters)
-    @extracts = extracts
+    @extracts = extracts.group_by(&:classification_id).map { |_, group| ExtractsForClassification.new(group) }
     @filters = filters.with_indifferent_access
   end
 
   def to_a
-    filter_by_extractor_ids(filter_by_subrange(extracts))
+    filter_by_extractor_ids(filter_by_subrange(filter_by_repeatedness(extracts))).flat_map(&:extracts)
   end
 
   private
 
-  def filter_by_subrange(extracts)
-    group_extracts(extracts)[subrange].flat_map { |i| i[:data] }
+  def filter_by_repeatedness(extracts)
+    case @filters[:repeated_classifications] || "keep_all"
```

Comment:
Should we default to `keep_first` instead? I think so, but it does break existing behaviour.